src/pv.rs

```diff
@@ -1,70 +1,73 @@
-#![allow(non_upper_case_globals)]
-#![allow(non_camel_case_types)]
-#![allow(non_snake_case)]
-#![allow(dead_code)]
-#![allow(clippy::useless_transmute)]
-#![allow(clippy::upper_case_acronyms)]
-include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+mod bindings {
+    #![allow(non_upper_case_globals)]
+    #![allow(non_camel_case_types)]
+    #![allow(non_snake_case)]
+    #![allow(dead_code)]
+    #![allow(clippy::all)]
+    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+}
+
+use bindings::*;
 
 use core::ffi::*;
 use pnet::datalink::{interfaces, NetworkInterface};
 use std::alloc::{alloc_zeroed, Layout};
 use std::ffi::CString;
+use std::ptr::copy;
 
 const DEFAULT_HEADROOM: u32 = 256;
 
-fn packet_dump(packet: &PvPacket) {
-    let chunk_address = packet.private as u64;
-    let buffer_address: *const u8 = packet.buffer.cast_const();
-
-    let length: u32 = packet.size;
-    let mut count: u32 = 0;
-
-    unsafe {
-        println!("---packet dump--- chunk addr: {}", chunk_address);
-
-        loop {
-            let read_offset: usize = count as usize;
-            let read_address: *const u8 = buffer_address.add(read_offset);
-            print!("{:02X?} ", std::ptr::read(read_address));
-
-            count += 1;
-            if count == length {
-                break;
-            } else if count % 8 == 0 {
-                print!(" ");
-                if count % 16 == 0 {
-                    println!();
-                }
-            }
-        }
-    }
-    println!("\n-------\n");
-}
-
 #[derive(Debug)]
-pub struct PvPacket {
+pub struct Packet {
     pub start: u32,       // payload offset pointing the start of payload. ex. 256
     pub end: u32,         // payload offset point the end of payload. ex. 1000
-    pub size: u32,        // total size of buffer. ex. 2048
+    pub buffer_size: u32, // total size of buffer. ex. 2048
     pub buffer: *mut u8,  // buffer address.
     private: *mut c_void, // DO NOT touch this.
 }
 
-impl PvPacket {
-    fn new() -> PvPacket {
-        PvPacket {
+impl Packet {
+    fn new() -> Packet {
+        Packet {
             start: 0,
             end: 0,
-            size: 0,
+            buffer_size: 0,
             buffer: std::ptr::null_mut(),
             private: std::ptr::null_mut(),
         }
     }
+
+    // replace payload with new data, and return Ok(remaining size) or Err(excess size) if failed
+    pub fn replace_data(&mut self, new_data: &Vec<u8>) -> Result<u32, u32> {
```

Comment:
Result type should be documented.
But I think the result value is not used anywhere, And the result value are nonsense

Reply:
Changed it with `Result<(), String>`